pkg/tsdb/phlare/instance.go

```diff
@@ -165,6 +170,17 @@ type BackendTypeRespBody struct {
 
 // backendType is a simplistic test to figure out if we are speaking to phlare or pyroscope backend
 func (d *PhlareDatasource) backendType(ctx context.Context, req *backend.CallResourceRequest, sender backend.CallResourceResponseSender) error {
+	// To prevent any user sending arbitrary URL for us to test with we allow this only for users who can edit the datasource
+	// as config page is where this is meant to be used.
+	ok, err := d.isUserAllowedToEditDatasource(ctx)
```

Comment:
Because when you create a data source and start filling in the fields it is not saved yet so you don't have the settings on the backend. We could save on every input change or onBlur but that would change how all config pages work which seems a bit invasive. Because of that I am not sure how to do this otherwise.

Reply:
So this datasource supports both phlare and pyroscope backends, correct? Very hard to understand the full picture giving this PR doesn't include the ConfigEditor changes. Depending on which backend selected you want to render certain fields in the config editor? 

In general, doing things like version detect must happen after a datasource have been saved. This is how graphite, postgres and Prometheus does it (I think). Does this imply you should keep phlare and pyroscope as separate datasources and the new one only supporting Grafana Pyroscope?

Reply:
Or is it that you just want to auto-detect the backend and set the backend type automatically as a nice-to-have UX? In that case would suggest looking into how the Prometheus datasource implemented their version detection.

Reply:
> very hard to understand the full picture giving this PR doesn't include the ConfigEditor changes.
I did not realize this issue only after the original PR was merged https://github.com/grafana/grafana/pull/66989.

> Or is it that you just want to auto-detect the backend and set the backend type automatically as a nice-to-have UX

Exactly and based on what I see in prometheus: 
https://github.com/grafana/grafana/blob/0a9240aebae687410a7d25b81fe9cd761fbf87da/public/app/plugins/datasource/prometheus/configuration/PromSettings.tsx#L98 the issue they mention is exactly what I wanted to avoid. As I mentioned saving the data source silently on some (or every?) option change when there is a separate `Save` button seemed like a less appealing option to me.

Reply:
Thanks I understand. To simplify I would suggest the user to select which backend to use. Imagine one of these  backends would go away eventually? Otherwise would do like Prometheus even though it's not optimal.

If we would implement/move this datasource as an external plugin or if we decide to run it not part of core Grafana (as its own process) this solution would fall apart why I don't recommend it. 

Security is another aspect which already had some eyes, good. But even though only users having access to configure datasources can call this endpoint and provide a URL, in Grafana Cloud we cannot trust any users even if they're admins. 

At the end up to you to decide.

Reply:
@marefr if you don't mind I would merge this now. The functionality itself is already in G10, so this just adds the auth which make it a bit better imho. Changing the way this works would be a bit more work, and we can change that later without breaking anything, and I think it needs some discussion about how to ideally do these kind of automatic config things.

Also I am still not understanding the " Grafana Cloud we cannot trust any users even if they're admins" and I heard it multiple times. I mean we trust them enough to create a data source with custom URL so I assume this is the same thing isn't it?

Reply:
Sorry for slow response. Been busy with other things. 

> if you don't mind I would merge this now. The functionality itself is already in G10, so this just adds the auth which make it a bit better imho. Changing the way this works would be a bit more work, and we can change that later without breaking anything, and I think it needs some discussion about how to ideally do these kind of automatic config things.

Sure, sounds good :+1: As I said earlier, at the end up to you to decide. But getting rid of that RBAC check within the plugin would be nice :) 

> Also I am still not understanding the " Grafana Cloud we cannot trust any users even if they're admins" and I heard it multiple times. I mean we trust them enough to create a data source with custom URL so I assume this is the same thing isn't it?

I guess I was under the impression the URL sent from client to the backend in your case was the whole URL including path, e.g. the URL path is not limited by the backend, but looking at the code again seems like you're specifically fetching "ProfileTypes" so should be similar as a regular HTTP datasource. 

In general, i's very easy to forget about this aspect and maybe taking certain security things for granted why I always try to put this out as a reminder :)